### PR TITLE
Fix providers created during updates with `--target` with uncreated parents

### LINKED ISSUE #12621: pulumi up target can cause providers in ComponentResources to become unparented
### What happened?

([context](https://pulumi.slack.com/archives/CKJ4C845D/p1678135981871219))

Running `pulumi up --target "<existing-urn>" --yes --skip-preview` on an existing stack that has had changes in code that would otherwise (without the `target`) create a new ComponentResource containing a Provider, causes the Provider to be created. This then throws `post-step event returned an error: failed to verify snapshot: child resource <urn:of::provider> refers to missing parent <urn:of::componentresource>`

Trying to then run `pulumi up` doesn't fix it: you get `pre-step event returned an error: failed to verify snapshot`. This leaves the parent-less Provider in the stack, so any further updates are blocked. You have to `stack export`, remove the provider, and `import`

### Expected Behavior

We'd expect that the `target` doesn't create resources that aren't being targeted.

### Steps to reproduce

`pulumi new` -> `aws-typescript`. All the defaults.
`pulumi config set aws:profile myprofile`
`pulumi config set aws:region us-west-2`
`pulumi config set other:profile otherprofile`
`pulumi config set other:region us-west-2`

Create a new file `component.ts` with:
```ts
import * as pulumi from '@pulumi/pulumi';
import * as aws from '@pulumi/aws';

export class OtherAccountBucket extends pulumi.ComponentResource {
    public readonly bucket: aws.s3.Bucket;
    constructor(opts: {}, componentOpts?: pulumi.ComponentResourceOptions) { 
        super('test:otheraccount:bucket', 'otheraccountbucket', {}, componentOpts);

        const otherAccount = new pulumi.Config('other');
        const otherAccountProvider = new aws.Provider(
            'other-account-provider',
            {
                profile: otherAccount.get('profile'),
                region: otherAccount.get('region') as pulumi.Input<aws.Region>
            },
            { 
                parent: this
            }
        );

        this.bucket = new aws.s3.Bucket(
            `testbucket`,
            { 
                bucket: `s3-testbucket`,
                acl: 'private'
            },
            {
                parent: this,
                provider: otherAccountProvider
            }
        );
    }
}
```
And modify `index.ts` to:
```ts
import * as pulumi from "@pulumi/pulumi";
import * as aws from "@pulumi/aws";
import { OtherAccountBucket } from "./component";

// Create an AWS resource (S3 Bucket)
const bucket = new aws.s3.Bucket("my-bucket");
// const otherBucket = new OtherAccountBucket("other-bucket");

// Export the name of the bucket
export const bucketName = bucket.id;
```

`pulumi up` stands up the first S3 `my-bucket` as intended. Get the bucket urn, uncomment the `otherBucket`, and run:

`pulumi up --target "urn:pulumi:dev::ProviderTest::aws:s3/bucket:Bucket::my-bucket" --yes --skip-preview`

Will generate the error:
```
Diagnostics:
  pulumi:providers:aws (other-account-provider):
    error: post-step event returned an error: failed to verify snapshot: child resource urn:pulumi:dev::ProviderTest::test:otheraccount:bucket$pulumi:providers:aws::other-account-provider refers to missing parent urn:pulumi:dev::ProviderTest::test:otheraccount:bucket::otheraccountbucket

  pulumi:pulumi:Stack (ProviderTest-dev):
    error: resource complete event returned an error: failed to verify snapshot: child resource urn:pulumi:dev::ProviderTest::test:otheraccount:bucket$pulumi:providers:aws::other-account-provider refers to missing parent urn:pulumi:dev::ProviderTest::test:otheraccount:bucket::otheraccountbucket
    error: update failed
```

### Output of `pulumi about`

CLI          
Version      3.61.1
Go Version   go1.20.3
Go Compiler  gc

Plugins
NAME    VERSION
aws     5.35.0
awsx    1.0.2
docker  3.6.1
nodejs  unknown

Host     
OS       darwin
Version  12.6.3
Arch     x86_64

This project is written in nodejs: executable='/usr/local/bin/node' version='v19.7.0'

TYPE                  URN
pulumi:pulumi:Stack   urn:pulumi:dev::ProviderTest::pulumi:pulumi:Stack::ProviderTest-dev
pulumi:providers:aws  urn:pulumi:dev::ProviderTest::pulumi:providers:aws::default_5_35_0
aws:s3/bucket:Bucket  urn:pulumi:dev::ProviderTest::aws:s3/bucket:Bucket::my-bucket
pulumi:providers:aws  urn:pulumi:dev::ProviderTest::test:otheraccount:bucket$pulumi:providers:aws::other-account-provider

Found no pending operations associated with dev

Backend        
Name           pulumi.com

Dependencies:
NAME            VERSION
@pulumi/aws     5.35.0
@pulumi/awsx    1.0.2
@pulumi/pulumi  3.61.1
@types/node     16.18.23

### Additional context

_No response_

### Contributing

Vote on this issue by adding a 👍 reaction. 
To contribute a fix for this issue, leave a comment (and link to your pull request, if you've opened one already).

### PR DESCRIPTION
**Pull request by dixler**
# Description

Provider resources are now skipped if not targeted for creation.

Fixes #12621 

## Checklist

- [x] I have added tests that prove my fix is effective or that my feature works

- [x] I have run `make changelog` and committed the `changelog/pending/<file>` documenting my change

- [ ] Yes, there are changes in this PR that warrants bumping the Pulumi Service API version

### CODE REVIEW

File: pkg/resource/deploy/step_generator.go [hunk 1]
```diff
@@ -849,7 +849,7 @@ func (sg *stepGenerator) generateSteps(event RegisterResourceEvent) ([]Step, res
 	// We will also not record this non-created resource into the checkpoint as it doesn't actually
 	// exist.
 
-	if !isTargeted && !providers.IsProviderType(goal.Type) {
+	if !isTargeted {
```
Comment (dixler) [inline pkg/resource/deploy/step_generator.go:852]: tl;dr

previously: non-targeted resources are skipped as long as they're not providers.
now: all non-targeted resources are skipped.

**Review by abhinav — COMMENTED** [review]
Sounds like a good change in theory and LGTM on that end, but if we can get someone familiar with step generator to also look at this, that would be great.

FWIW, I went spelunking and this special case has been present since #3426.

File: pkg/resource/deploy/step_generator.go [hunk 2]
```diff
@@ -611,9 +611,14 @@ func (sg *stepGenerator) generateSteps(event RegisterResourceEvent) ([]Step, res
...
 
+	isTargeted := sg.isTargetedForUpdate(new)
+	if isTargeted {
+		sg.updateTargetsOpt.addLiteral(urn)
+	}
+
 	// Ensure the provider is okay with this resource and fetch the inputs to pass to subsequent methods.
 	var err error
-	if prov != nil {
+	if isTargeted && prov != nil {
```
Comment (Frassle) [inline pkg/resource/deploy/step_generator.go:621]: We've made a very similar change to this before (https://github.com/pulumi/pulumi/pull/9960) which triggered a pretty bad P0 (https://github.com/pulumi/pulumi/issues/10020).

I'd be pretty hesitant to make changes like this again unless we think we have tests to explain and cover what happened back then (https://github.com/pulumi/pulumi/issues/10056).

Comment (dixler) [reply pkg/resource/deploy/step_generator.go:621]: I'm looking at the code you linked, and it had some early exits. This change makes the branch more selective and doesn't call prov.Check() instead.

I'll add a test to add coverage for this issue https://github.com/pulumi/pulumi/issues/10056

File: pkg/engine/lifecycletest/update_plan_test.go [hunk 3]
```diff
@@ -1640,3 +1640,289 @@ func TestPlannedUpdateWithDependentDelete(t *testing.T) {
...
+			UpdateTargets: deploy.NewUrnTargets([]string{
+				"urn:pulumi:test::test::pkgA:m:typA::resB",
+				"urn:pulumi:test::test::pulumi:providers:pkgA::default",
+			}),
+		}, false, p.BackendClient, nil)
+		assert.Nil(t, res)
+	}
+}
+
+func TestTargetedDoesNotCreateProvider(t *testing.T) {
```
Comment (dixler) [inline pkg/engine/lifecycletest/update_plan_test.go:1857]: This test captures the change in behavior (i.e. fails on master and succeeds on this branch)

File: pkg/resource/deploy/step_generator.go (hunk 2 above)
Comment (dixler) [reply pkg/resource/deploy/step_generator.go:617]: The issue the user ran into was very vague and I cannot reproduce it in a reasonable amount of time.

File: pkg/resource/deploy/step_generator.go [hunk 4]
```diff
@@ -298,11 +298,10 @@ func (sg *stepGenerator) GenerateSteps(event RegisterResourceEvent) ([]Step, res
 			if !isAliased {
 				urn = s.URN()
 			}
-			resourcePlan, ok := sg.deployment.newPlans.get(urn)
-			if !ok {
-				return nil, result.Errorf("Expected a new resource plan for %v", urn)
+			if resourcePlan, ok := sg.deployment.newPlans.get(urn); ok {
+				// If the resource is in the plan, add the operation to the plan.
+				resourcePlan.Ops = append(resourcePlan.Ops, s.Op())
```
Comment (dixler) [inline pkg/resource/deploy/step_generator.go:303]: to gate on isTargeted below means that urns are not added to this. This is caught above by checking that unspecified resources are sames.

File: pkg/engine/lifecycletest/update_plan_test.go [hunk 5]
```diff
@@ -1640,3 +1640,136 @@ func TestPlannedUpdateWithDependentDelete(t *testing.T) {
...
+
+		host := deploytest.NewPluginHost(nil, nil, program, loaders...)
+		return host
+	}()
+
+	p := &TestPlan{}
+
+	project := p.GetProject()
+
+	// Define a function that creates new resource plans as resource plans get mutated during use.
```
Comment (Frassle) [inline pkg/engine/lifecycletest/update_plan_test.go:1683]: You can just clone the plan before passing it into the next update, e.g see https://github.com/pulumi/pulumi/pull/12628/files#diff-114d9ebc788ae0f44bb57d805a87fed37c05998882572331967f95f6ab763afeR1638

File: pkg/engine/lifecycletest/update_plan_test.go [hunk 6]
```diff
@@ -1640,3 +1640,136 @@ func TestPlannedUpdateWithDependentDelete(t *testing.T) {
...
+			}),
+		}, p.BackendClient, nil)
+		assert.Nil(t, res)
+		assert.NotNil(t, plan)
+		return plan
+	}
+
+	// Check that running an update with everything targeted fails due to our plan being constrained
+	// to the resource and default provider.
+	{
```
Comment (Frassle) [inline pkg/engine/lifecycletest/update_plan_test.go:1701]: We don't generally indent test parts in blocks like this

File: pkg/engine/lifecycletest/update_plan_test.go (hunk 3 above)
Comment (Frassle) [reply pkg/engine/lifecycletest/update_plan_test.go:1726]: This test isn't using updatePlans and so shouldn't be in update_plan_test.go

File: pkg/engine/lifecycletest/update_plan_test.go [hunk 7]
```diff
@@ -1640,3 +1640,136 @@ func TestPlannedUpdateWithDependentDelete(t *testing.T) {
...
+
+	p := &TestPlan{}
+
+	project := p.GetProject()
+
+	// Check that update fails because the default provider is not created (as it is not targeted).
+	{
+		_, res := TestOp(Update).Run(project, p.GetTarget(t, nil), UpdateOptions{
+			Host:         host,
+			Experimental: true,
```
Comment (Frassle) [inline pkg/engine/lifecycletest/update_plan_test.go:1755]: No need to set Experimental

File: pkg/engine/lifecycletest/update_plan_test.go [hunk 8]
```diff
@@ -1640,3 +1640,136 @@ func TestPlannedUpdateWithDependentDelete(t *testing.T) {
...
+
+		host := deploytest.NewPluginHost(nil, nil, program, loaders...)
+		return host
+	}()
+
+	p := &TestPlan{}
+
+	project := p.GetProject()
+
+	// Check that update fails because the default provider is not created (as it is not targeted).
```
Comment (Frassle) [inline pkg/engine/lifecycletest/update_plan_test.go:1751]: There's part of me a little worried about this. We don't generally make users think about default providers too much and this is going to require them to learn default provider URNs just to create new targetted resources.

I think the change "explicit provider resources shouldn't be created unless targeted" is fine, as even if people are currently relying on this behavior I'm much happier pointing out to them that this is like having a normal resource dependency and expecting it to work if you only targeted the dependent resource.

File: pkg/engine/lifecycletest/target_test.go [hunk 9]
```diff
@@ -914,3 +970,50 @@ func newResource(urn, parent resource.URN, id resource.ID, provider string, depe
 		Parent:               parent,
 	}
 }
+
+// TestTargetedDoesNotCreateProvider checks that an update that targets a resource does not create the default
```
Comment (Frassle) [inline pkg/engine/lifecycletest/target_test.go:974]: We said we'd change this so default providers don't need to be targetted.

File: pkg/engine/lifecycletest/update_plan_test.go [hunk 10]
```diff
@@ -1640,3 +1640,79 @@ func TestPlannedUpdateWithDependentDelete(t *testing.T) {
...
+	project := p.GetProject()
+
+	// Create the update plan with only targeted resources.
+	plan, res := TestOp(Update).Plan(project, p.GetTarget(t, nil), UpdateOptions{
+		Host:         host,
+		Experimental: true,
+		GeneratePlan: true,
+		UpdateTargets: deploy.NewUrnTargets([]string{
+			"urn:pulumi:test::test::pkgA:m:typA::resB",
+			"urn:pulumi:test::test::pulumi:providers:pkgA::default",
```
Comment (Frassle) [inline pkg/engine/lifecycletest/update_plan_test.go:1690]: Don't want to require default providers on --targets

**Review by Frassle — APPROVED** [review]
Tests look good.

**Comment by dixler** [comment]
bors merge

**Pull request merged**